File: webapp/src/egp/process.ts
/**
 * Simulation of the FIR process and its uniform quantization — the port of
 * egp/process.py. Innovations W_{2−L} … W_n are drawn i.i.d. standard normal
 * so that every returned sample has its full history; there is no burn-in
 * transient, matching the "exact prior" initialization of the filter.
 */
import type { Rng } from './rng'

/** One stationary quantized draw y_t = round(x_t / delta), t = 1 … n. */
export function generateQuantized(taps: Float64Array, n: number, delta: number, rng: Rng): Int32Array {
  const L = taps.length
  const w = new Float64Array(n + L - 1)
  for (let i = 0; i < w.length; i++) w[i] = rng.normal()
  const y = new Int32Array(n)
  for (let t = 0; t < n; t++) {
    let x = 0
    // x_t = sum_j taps[j] * W_{t-j}; w is indexed so w[t + L - 1] = W_t.
    for (let j = 0; j < L; j++) x += taps[j] * w[t + L - 1 - j]
    y[t] = Math.floor(x / delta + 0.5)
  }
  return y
}

/** Derive an independent seed for replicate `index` from a base seed. */
export function replicateSeed(base: number, index: number): number {
  let z = (base ^ Math.imul(index + 1, 0x9e3779b9)) >>> 0
  z = Math.imul(z ^ (z >>> 16), 0x45d9f3b) >>> 0
  z = Math.imul(z ^ (z >>> 16), 0x45d9f3b) >>> 0
  return (z ^ (z >>> 16)) >>> 0
}

File: egp/src/egp/process.py
"""Simulation of the process and its uniform quantization."""

from __future__ import annotations

import numpy as np
from scipy.signal import fftconvolve


def quantize(x, delta: float) -> np.ndarray:
    """Round to the nearest multiple of ``delta``; returns integer symbols."""
    return np.floor(np.asarray(x, dtype=float) / float(delta) + 0.5).astype(np.int64)


def dequantize(y, delta: float) -> np.ndarray:
    return np.asarray(y, dtype=float) * float(delta)


def generate(taps, n: int, rng: np.random.Generator) -> np.ndarray:
    """Draw a stationary sample X_1..X_n of the FIR process.

    Innovations W_{2-L} .. W_n are drawn i.i.d. standard normal so that every
    returned sample has its full history; there is no burn-in transient.
    """
    taps = np.asarray(taps, dtype=float).ravel()
    n_taps = taps.size
    w = rng.standard_normal(n + n_taps - 1)
    if n_taps == 1:
        return taps[0] * w
    return fftconvolve(w, taps, mode="valid")[:n]


def generate_quantized(taps, n: int, delta: float, rng: np.random.Generator):
    """Return ``(x, y)``: the continuous sample and its integer symbols."""
    x = generate(taps, n, rng)
    return x, quantize(x, delta)
